File: backend/app/schemas/checkout.py
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


CheckoutThemeMode = Literal["light", "dark"]
CheckoutPaymentMethod = Literal["pix", "card"]
CheckoutDiscountType = Literal["fixed", "percent"]


class CheckoutAppearanceIn(BaseModel):
    key: str = Field(..., min_length=1, max_length=120)
    name: str = Field(..., min_length=1, max_length=255)
    theme_mode: CheckoutThemeMode = "dark"
    desktop_image_url: str | None = None
    mobile_banner_url: str | None = None
    active: bool = True

    @field_validator("key")
    @classmethod
    def normalize_key(cls, value: str) -> str:
        return value.strip().lower()


class CheckoutAppearanceOut(CheckoutAppearanceIn):
    pass


class CheckoutOfferIn(BaseModel):
    key: str = Field(..., min_length=1, max_length=120)
    title: str = Field(..., min_length=1, max_length=255)
    footer_product_label: str = Field(default="Plano profissional", min_length=1, max_length=120)
    plan_key: str = Field(default="professional", min_length=1, max_length=50)
    billing_cycle: Literal["monthly", "annual"] = "monthly"
    amount: Decimal = Field(default=Decimal("49.90"), gt=0)
    active: bool = True
    subtitle: str | None = None
    checkout_key: str | None = Field(default=None, max_length=120)

    @field_validator("key")
    @classmethod
    def normalize_key(cls, value: str) -> str:
        return value.strip().lower()

    @field_validator("checkout_key")
    @classmethod
    def normalize_checkout_key(cls, value: str | None) -> str | None:
        normalized = (value or "").strip().lower()
        return normalized or None


class CheckoutOfferOut(CheckoutOfferIn):
    amount: Decimal


class CheckoutCouponIn(BaseModel):
    code: str = Field(..., min_length=1, max_length=80)
    title: str = Field(..., min_length=1, max_length=255)
    discount_type: CheckoutDiscountType = "fixed"
    value: Decimal = Field(..., gt=0)
    offer_keys: list[str] = Field(default_factory=list)
    active: bool = True

    @field_validator("code")
    @classmethod
    def normalize_code(cls, value: str) -> str:
        return value.strip().upper()

    @field_validator("offer_keys")
    @classmethod
    def normalize_offer_keys(cls, values: list[str]) -> list[str]:
        return [value.strip().lower() for value in values if value and value.strip()]


class CheckoutCouponOut(CheckoutCouponIn):
    value: Decimal
    usage_count: int = 0


class CheckoutSettingsUpdate(BaseModel):
    is_active: bool = True
    offers: list[CheckoutOfferIn] = Field(default_factory=list)
    coupons: list[CheckoutCouponIn] = Field(default_factory=list)
    checkouts: list[CheckoutAppearanceIn] = Field(default_factory=list)


class CheckoutSettingsOut(BaseModel):
    id: int
    is_active: bool
    offers: list[CheckoutOfferOut] = Field(default_factory=list)
    coupons: list[CheckoutCouponOut] = Field(default_factory=list)
    checkouts: list[CheckoutAppearanceOut] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class CheckoutPublicConfigOut(BaseModel):
    theme_mode: CheckoutThemeMode
    desktop_image_url: str | None = None
    mobile_banner_url: str | None = None
    offer: CheckoutOfferOut


class CheckoutSessionCreate(BaseModel):
    offer_key: str = Field(..., min_length=1, max_length=120)
    customer_name: str = Field(..., min_length=1, max_length=255)
    customer_email: str = Field(..., min_length=3, max_length=255)
    customer_document: str = Field(..., min_length=11, max_length=18)
    customer_phone: str = Field(..., min_length=10, max_length=20)
    customer_zipcode: str = Field(..., min_length=8, max_length=10)
    coupon_code: str | None = Field(default=None, max_length=80)
    metadata: dict[str, Any] | None = None

    @field_validator("offer_key")
    @classmethod
    def normalize_offer_key(cls, value: str) -> str:
        return value.strip().lower()

    @field_validator("coupon_code")
    @classmethod
    def normalize_coupon_code(cls, value: str | None) -> str | None:
        normalized = (value or "").strip().upper()
        return normalized or None

    @field_validator("customer_email")
    @classmethod
    def normalize_email(cls, value: str) -> str:
        return value.strip().lower()

    @field_validator("customer_document")
    @classmethod
    def validate_document(cls, value: str) -> str:
        digits = "".join(ch for ch in value if ch.isdigit())
        if len(digits) not in {11, 14}:
            raise ValueError("customer_document must have 11 (CPF) or 14 (CNPJ) digits")
        return value


class CheckoutUpgradeSessionDetailsUpdate(BaseModel):
    customer_name: str = Field(..., min_length=1, max_length=255)
    customer_email: str = Field(..., min_length=3, max_length=255)
    customer_document: str = Field(..., min_length=11, max_length=18)
    customer_phone: str = Field(..., min_length=10, max_length=20)
    customer_zipcode: str = Field(..., min_length=8, max_length=10)

    @field_validator("customer_document")
    @classmethod
    def validate_document(cls, value: str) -> str:
        digits = "".join(ch for ch in value if ch.isdigit())
        if len(digits) not in {11, 14}:
            raise ValueError("customer_document must have 11 (CPF) or 14 (CNPJ) digits")
        return value


class CheckoutCouponPreviewIn(BaseModel):
    offer_key: str = Field(..., min_length=1, max_length=120)
    coupon_code: str = Field(..., min_length=1, max_length=80)

    @field_validator("offer_key")
    @classmethod
    def normalize_offer_key(cls, value: str) -> str:
        return value.strip().lower()

    @field_validator("coupon_code")
    @classmethod
    def normalize_coupon_code(cls, value: str) -> str:
        return value.strip().upper()


class CheckoutCouponPreviewOut(BaseModel):
    offer_key: str
    amount: Decimal
    original_amount: Decimal
    discount_amount: Decimal
    applied_coupon_code: str


class CheckoutSessionOut(BaseModel):
    token: str
    offer_key: str
    product_name: str
    billing_cycle: str
    amount: Decimal
    original_amount: Decimal | None = None
    discount_amount: Decimal | None = None
    applied_coupon_code: str | None = None
    status: str
    payment_method: str | None = None
    pix_mode: Literal["automatic", "conventional"] | None = None
    customer_name: str
    customer_email: str
    customer_document: str
    customer_phone: str
    customer_zipcode: str
    pix_copy_paste: str | None = None
    pix_qr_code_base64: str | None = None
    pix_expiration_date: datetime | None = None
    paid_at: datetime | None = None
    password_defined_at: datetime | None = None
    user_exists: bool = False
    requires_password_setup: bool = True
    is_upgrade: bool = False
    locked_profile_fields: dict[str, bool] | None = None


class CheckoutPixStartRequest(BaseModel):
    token: str


class CheckoutCardStartRequest(BaseModel):
    token: str
    holder_name: str = Field(..., min_length=1, max_length=255)
    card_number: str = Field(..., min_length=13, max_length=24)
    expiry_month: str = Field(..., min_length=2, max_length=2)
    expiry_year: str = Field(..., min_length=2, max_length=4)
    ccv: str = Field(..., min_length=3, max_length=4)
    installment_count: int = Field(default=1, ge=1, le=12)
    remote_ip: str | None = None


class CheckoutPasswordRequest(BaseModel):
    token: str
    password: str = Field(..., min_length=8, max_length=120)


class CheckoutPasswordResponse(BaseModel):
    detail: str
    email: str


class CheckoutTrackEventIn(BaseModel):
    event_name: str = Field(..., min_length=2, max_length=80)
    step: str | None = Field(default=None, max_length=40)
    status: str | None = Field(default=None, max_length=40)
    payment_method: str | None = Field(default=None, max_length=30)
    duration_ms: int | None = Field(default=None, ge=0)
    error_message: str | None = Field(default=None, max_length=2000)
    metadata: dict[str, Any] | None = None


class CheckoutTrackEventOut(BaseModel):
    ok: bool = True


class CheckoutTrackingItemOut(BaseModel):
    id: int
    token: str | None = None
    offer_key: str
    customer_name: str
    customer_email: str
    customer_document: str | None = None
    customer_phone: str
    customer_zipcode: str
    payment_method_selected: str | None = None
    signed_at: datetime | None = None
    password_defined_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    events_count: int = 0
    success_events_count: int = 0
    error_events_count: int = 0
    last_event_name: str | None = None
    last_event_at: datetime | None = None
    last_ip_address: str | None = None
    last_ip_city: str | None = None
    last_ip_region: str | None = None
    last_ip_country: str | None = None
    total_time_seconds: int | None = None
    sessions_count: int = 1


class CheckoutTrackingEventOut(BaseModel):
    id: int
    token: str
    offer_key: str
    event_name: str
    step: str | None = None
    status: str | None = None
    payment_method: str | None = None
    duration_ms: int | None = None
    error_message: str | None = None
    ip_address: str | None = None
    ip_country: str | None = None
    ip_region: str | None = None
    ip_city: str | None = None
    created_at: datetime


class CheckoutTrackingDocumentDetailOut(BaseModel):
    customer_document: str
    customer_name: str
    customer_email: str
    customer_phone: str
    sessions_count: int
    events_count: int
    error_events_count: int
    signed_count: int
    password_defined_count: int
    first_seen_at: datetime | None = None
    last_seen_at: datetime | None = None
    events: list[CheckoutTrackingEventOut] = Field(default_factory=list)


class CheckoutOfferReportItemOut(BaseModel):
    offer_key: str
    offer_title: str
    checkout_key: str | None = None
    is_active: bool = True
    signed_count: int = 0
    upgrade_count: int = 0
    direct_count: int = 0
    total_count: int = 0
    last_signed_at: datetime | None = None

File: frontend/src/services/checkout.ts
import api from "./api";

export type CheckoutThemeMode = "light" | "dark";
export type CheckoutDiscountType = "fixed" | "percent";

export interface CheckoutAppearance {
  key: string;
  name: string;
  theme_mode: CheckoutThemeMode;
  desktop_image_url?: string | null;
  mobile_banner_url?: string | null;
  active: boolean;
}

export interface CheckoutOffer {
  key: string;
  title: string;
  footer_product_label: string;
  plan_key: string;
  billing_cycle: "monthly" | "annual";
  amount: string;
  active: boolean;
  subtitle?: string | null;
  checkout_key?: string | null;
}

export interface CheckoutCoupon {
  code: string;
  title: string;
  discount_type: CheckoutDiscountType;
  value: string;
  offer_keys: string[];
  active: boolean;
  usage_count?: number;
}

export interface CheckoutSettings {
  id: number;
  is_active: boolean;
  offers: CheckoutOffer[];
  coupons: CheckoutCoupon[];
  checkouts: CheckoutAppearance[];
  created_at: string;
  updated_at: string;
}

export interface CheckoutPublicConfig {
  theme_mode: CheckoutThemeMode;
  desktop_image_url?: string | null;
  mobile_banner_url?: string | null;
  offer: CheckoutOffer;
}

export interface CheckoutSessionPayload {
  offer_key: string;
  customer_name: string;
  customer_email: string;
  customer_document: string;
  customer_phone: string;
  customer_zipcode: string;
  coupon_code?: string | null;
  metadata?: Record<string, unknown>;
}

export interface CheckoutSession {
  token: string;
  offer_key: string;
  product_name: string;
  billing_cycle: string;
  amount: string;
  original_amount?: string | null;
  discount_amount?: string | null;
  applied_coupon_code?: string | null;
  status: string;
  payment_method?: string | null;
  pix_mode?: "automatic" | "conventional" | null;
  customer_name: string;
  customer_email: string;
  customer_document: string;
  customer_phone: string;
  customer_zipcode: string;
  pix_copy_paste?: string | null;
  pix_qr_code_base64?: string | null;
  pix_expiration_date?: string | null;
  paid_at?: string | null;
  password_defined_at?: string | null;
  user_exists: boolean;
  requires_password_setup: boolean;
  is_upgrade?: boolean;
  locked_profile_fields?: Record<string, boolean> | null;
}

export interface CheckoutCouponPreviewPayload {
  offer_key: string;
  coupon_code: string;
}

export interface CheckoutCouponPreview {
  offer_key: string;
  amount: string;
  original_amount: string;
  discount_amount: string;
  applied_coupon_code: string;
}

export interface CheckoutCardPayload {
  token: string;
  holder_name: string;
  card_number: string;
  expiry_month: string;
  expiry_year: string;
  ccv: string;
  installment_count: number;
  remote_ip?: string | null;
}

export interface CheckoutTrackPayload {
  event_name: string;
  step?: string | null;
  status?: string | null;
  payment_method?: string | null;
  duration_ms?: number | null;
  error_message?: string | null;
  metadata?: Record<string, unknown> | null;
}

export interface CheckoutTrackingItem {
  id: number;
  token: string;
  offer_key: string;
  customer_name: string;
  customer_email: string;
  customer_document?: string | null;
  customer_phone: string;
  customer_zipcode: string;
  payment_method_selected?: string | null;
  signed_at?: string | null;
  password_defined_at?: string | null;
  created_at: string;
  updated_at: string;
  events_count: number;
  success_events_count: number;
  error_events_count: number;
  last_event_name?: string | null;
  last_event_at?: string | null;
  last_ip_address?: string | null;
  last_ip_city?: string | null;
  last_ip_region?: string | null;
  last_ip_country?: string | null;
  total_time_seconds?: number | null;
  sessions_count?: number;
}

export interface CheckoutTrackingEventItem {
  id: number;
  token: string;
  offer_key: string;
  event_name: string;
  step?: string | null;
  status?: string | null;
  payment_method?: string | null;
  duration_ms?: number | null;
  error_message?: string | null;
  ip_address?: string | null;
  ip_country?: string | null;
  ip_region?: string | null;
  ip_city?: string | null;
  created_at: string;
}

export interface CheckoutTrackingDocumentDetail {
  customer_document: string;
  customer_name: string;
  customer_email: string;
  customer_phone: string;
  sessions_count: number;
  events_count: number;
  error_events_count: number;
  signed_count: number;
  password_defined_count: number;
  first_seen_at?: string | null;
  last_seen_at?: string | null;
  events: CheckoutTrackingEventItem[];
}

export interface CheckoutOfferReportItem {
  offer_key: string;
  offer_title: string;
  checkout_key?: string | null;
  is_active: boolean;
  signed_count: number;
  upgrade_count: number;
  direct_count: number;
  total_count: number;
  last_signed_at?: string | null;
}

export const getAdminCheckoutSettings = async () => {
  const response = await api.get<CheckoutSettings>("/admin/checkout-settings");
  return response.data;
};

export const saveAdminCheckoutSettings = async (payload: Omit<CheckoutSettings, "id" | "created_at" | "updated_at">) => {
  const response = await api.put<CheckoutSettings>("/admin/checkout-settings", payload);
  return response.data;
};

export const getCheckoutConfig = async (offerKey: string) => {
  const response = await api.get<CheckoutPublicConfig>(`/checkout/public/config/${offerKey}`);
  return response.data;
};

export const createCheckoutSession = async (payload: CheckoutSessionPayload) => {
  const response = await api.post<CheckoutSession>("/checkout/public/session", payload);
  return response.data;
};

export const createUpgradeCheckoutSession = async (offerKey: string, couponCode?: string | null) => {
  const response = await api.post<CheckoutSession>("/checkout/public/session/upgrade", {
    offer_key: offerKey,
    coupon_code: couponCode || null,
  });
  return response.data;
};

export const updateUpgradeCheckoutSessionDetails = async (
  token: string,
  payload: Pick<
    CheckoutSessionPayload,
    "customer_name" | "customer_email" | "customer_document" | "customer_phone" | "customer_zipcode"
  >
) => {
  const response = await api.put<CheckoutSession>(`/checkout/public/session/${encodeURIComponent(token)}/upgrade-details`, payload);
  return response.data;
};

export const previewCheckoutCoupon = async (payload: CheckoutCouponPreviewPayload) => {
  const response = await api.post<CheckoutCouponPreview>("/checkout/public/coupon-preview", payload);
  return response.data;
};

export const getCheckoutSession = async (token: string) => {
  const response = await api.get<CheckoutSession>(`/checkout/public/session/${token}`);
  return response.data;
};

export const refreshCheckoutSession = async (token: string) => {
  const response = await api.post<CheckoutSession>(`/checkout/public/session/${token}/refresh`);
  return response.data;
};

export const startPixCheckout = async (token: string) => {
  const response = await api.post<CheckoutSession>(`/checkout/public/session/${token}/pix`);
  return response.data;
};

export const startCardCheckout = async (token: string, payload: CheckoutCardPayload) => {
  const response = await api.post<CheckoutSession>(`/checkout/public/session/${token}/card`, payload);
  return response.data;
};

export const finishCheckoutPassword = async (token: string, password: string) => {
  const response = await api.post<{ detail: string; email: string }>(`/checkout/public/session/${token}/password`, {
    token,
    password,
  });
  return response.data;
};

export const trackCheckoutEvent = async (token: string, payload: CheckoutTrackPayload) => {
  await api.post<{ ok: boolean }>(`/checkout/public/session/${token}/track`, payload);
};

export const getAdminCheckoutTracking = async (offerKey?: string, limit = 200) => {
  const response = await api.get<CheckoutTrackingItem[]>("/admin/checkout-settings/tracking", {
    params: { offer_key: offerKey || undefined, limit },
  });
  return response.data;
};

export const getAdminCheckoutTrackingDocument = async (customerDocument: string, offerKey?: string) => {
  const response = await api.get<CheckoutTrackingDocumentDetail>(`/admin/checkout-settings/tracking/document/${encodeURIComponent(customerDocument)}`, {
    params: { offer_key: offerKey || undefined },
  });
  return response.data;
};

export const getAdminCheckoutOfferReports = async (params?: { startDate?: string | null; endDate?: string | null }) => {
  const response = await api.get<CheckoutOfferReportItem[]>("/admin/checkout-settings/reports/offers", {
    params: {
      start_date: params?.startDate || undefined,
      end_date: params?.endDate || undefined,
    },
  });
  return response.data;
};
